utility/validator.go

```diff
@@ -1,24 +1,77 @@
 package utility
 
+// Internal business logic specific to validator behaviour and interactions.
+
 import (
 	"math/big"
 
 	coreTypes "github.com/pokt-network/pocket/shared/core/types"
 	typesUtil "github.com/pokt-network/pocket/utility/types"
 )
 
+// handleByzantineValidators identifies & handles byzantine or faulty validators.
+// This includes validators who double signed, didn't sign at all or disagree with 2/3+ majority.
+// IMPROVE: Need to add more logging to this function.
+// INCOMPLETE: handleByzantineValidators is a WIP and needs to be fully designed, implemented, tested and documented
+func (u *utilityContext) handleByzantineValidators(prevBlockByzantineValidators [][]byte) typesUtil.Error {
+	maxMissedBlocks, err := u.getValidatorMaxMissedBlocks()
+	if err != nil {
+		return err
+	}
+
+	for _, address := range prevBlockByzantineValidators {
+		// Get the latest number of missed blocks by the validator
+		numMissedBlocks, err := u.store.GetValidatorMissedBlocks(address, u.height)
+		if err != nil {
+			return typesUtil.ErrGetMissedBlocks(err)
+		}
+
+		// increment missed blocks
+		numMissedBlocks++
+
+		// handle if under the threshold of max missed blocks
+		if numMissedBlocks < maxMissedBlocks {
+			if err := u.store.SetValidatorMissedBlocks(address, numMissedBlocks); err != nil {
+				return typesUtil.ErrSetMissedBlocks(err)
+			}
+			continue
+		}
+
+		// pause the validator for exceeding the threshold: numMissedBlocks >= maxMissedBlocks
+		if err := u.store.SetValidatorPauseHeight(address, u.height); err != nil {
+			return typesUtil.ErrSetPauseHeight(err)
+		}
+		// update the number of blocks it missed
+		if err := u.store.SetValidatorMissedBlocks(address, numMissedBlocks); err != nil {
+			return typesUtil.ErrSetMissedBlocks(err)
+		}
+		// burn validator for missing blocks
+		if err := u.burnValidator(address); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// burnValidator burns a validator's stake based on governance parameters for missing blocks
+// and begins unstaking if the stake falls below the necessary threshold
+// REFACTOR: Extend this to support burning other actors types & pools once the logic is implemented
 // ADDTEST: There are no good tests for this functionality, which MUST be added.
-func (u *utilityContext) burnValidator(burnPercent int, addr []byte) typesUtil.Error {
-	// TODO: Will need to extend this to support burning from other actors types & pools when the logic is implemented
-	validatorActorType := coreTypes.ActorType_ACTOR_TYPE_VAL
-	validatorPool := coreTypes.Pools_POOLS_VALIDATOR_STAKE
+func (u *utilityContext) burnValidator(addr []byte) typesUtil.Error {
```

Comment:
Confusingly, IMHO, this returns `Error` but it has been removed elsewhere

Related: https://github.com/pokt-network/pocket/pull/550/files#r1118743922